Add FLIC Dataset

WilliamHYZhang opened the pull request:
Hi,

I'm currently working on adding the FLIC Dataset (https://bensapp.github.io/flic-dataset.html).

Currently, the plan is to have the annotations as closely resemble the original examples provided in the examples.mat file as we can.

Here is a basic outline of what I did already:
- citations, description, urls, all loaded
- example.mat file is loaded with scipy.io (will change import to tfds.core.lazy_imports in a future commit) (note: example.mat file is analyzed as a multidimensional array, e.g. data["example"][index_of_example_image][feature], where feature corresponds with the values in the FeatureDict)
- established FeatureDict in _info with features corresponding to the values given by .dtype (attribute names, mostly either tdfs.features.Image(), tfds.features.Text(), or tfds.features.Tensor() with corresponding shapes, types such as tf.uint16 and tf.float64 have also been found and implemented)
- established train/test splits in _split_generators and passed parameters to _generate_examples, done in the form of two lists (test and train) containing indexes of examples in data (note: no validation split from what I observed, booleans for istest/istrain are never both true or false)
- have a skeletal _generate_example function that simply yield the index of the attributes in data for each example in split

Some issues (most important to least important)
- download_and_prepare is failing as attributes with type tfds.features.Sequence are unable be yielded in the _generate_example function, returns error that the shapes must have the same length (note: these attributes are "poselet_hit_idx" and "coords", both of which are np arrays that have variable length for each example, see lines 99 and 101.)
- the "idx" values in the "poselet_hit_idx" array are nowhere in the range [1, 29] as the name-key mapping might suggest, I have absolutely no clue whatsoever what these numbers are supposed to represent
- "torsobox" attribute is in form of of 4 float Tensor instead of BBox, don't know if should change, if yes, then also need to implement normalize_bbox function and include PIL library (also probably from tfds.core.lazy_imports) to determine image sizes

Thanks, look forward to continuing with this project and contributing to Tensorflow.

WilliamHYZhang commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,106 @@
+        "image": os.path.join(extract_path, "FLIC", "images", data["examples"][index][3]),
+        "poselet_hit_idx": {"idx": list(data["examples"][index][0])},
+        "moviename": data["examples"][index][1],
+        "coords": {"coord": list(data["examples"][index][2])},
```
This line should be rewritten to have "coords" return a dict of "x" and "y" tf.float64's. How should i use the zip() function for this?

WilliamHYZhang replied:
Disregard, fixed with push

WilliamHYZhang commented:
update: pushed with the fix, see output .json file [here](https://gist.github.com/WilliamHYZhang/a91ba59e867acfc00b8d773dae257938)

captain-pool commented:
Oops, I actually forgot to click Start Review Button. You should have got this review much earlier.😅

WilliamHYZhang commented:
@captain-pool Haha, thanks for the feedback, already fixed the issue with the coords, working on the updated split.

WilliamHYZhang commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,105 @@
+      if data["examples"][i][7]:
+        train_list.append(i)
+      elif data["examples"][i][8]:
+        test_list.append(i)
```
@captain-pool thanks for the suggestion, pushed update with fix.

captain-pool commented:
Issue reference: #547

WilliamHYZhang commented:
@captain-pool thank you for the reviews! Will work on implementing changes and then testing later today.

WilliamHYZhang commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,101 @@
+          "istest": True,
+        },
+      ),
+    ]
```
I originally had a split for validation, however I believe in the annotations that none of the examples have both is true false and is test false, one of them is always true. This leads me to think that there was only intended to be too splits for the data.

captain-pool replied:
> 
> 
> I originally had a split for validation, however I believe in the annotations that none of the examples have both is true false and is test false, one of them is always true. This leads me to think that there was only intended to be too splits for the data.

I see.

captain-pool commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,101 @@
+        builder=self,
+        description=_DESCRIPTION,
+        features=tfds.features.FeaturesDict({
+            "image": tfds.features.Image(),
```
As far as I remember the shape of all the images are more or less same. So, you can put that shape here as:
`tfds.features.Image(shape=(height, width, 3))`

WilliamHYZhang commented:
@captain-pool pushed new update with updated main file, new testing and fake examples also added. All tests are passing. Thanks so much!

captain-pool commented:
@cyfra, @rsepassi please take a look

captain-pool commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,95 @@
+set aside 20% (1016 images) of the data for testing. 
+"""
+
+_URL = "https://drive.google.com/uc?id=0B4K3PZp8xXDJN0Fpb0piVjQ3Y3M&export=download"
```
Hey, I just suddenly remembered there are two versions of this right? One full dataset and another partial? And this is the partial right?

Will it be possible for you to write a heavy configuration to provide both versions of the datasets?
For example: "flic/full" and "flic/small".

WilliamHYZhang replied:
Of course! On it.

WilliamHYZhang commented:
@captain-pool All done! Implemented the builderconfig and modified the testing and fake data for the new structure. All tests passing.

WilliamHYZhang commented:
Is there anything I should do before we can merge this and my other approved PRs in this repo? Thanks!

cyfra commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,132 @@
+    if data not in _DATA_OPTIONS:
+      raise ValueError("data must be one of %s" % _DATA_OPTIONS)
+
+    name = kwargs.get("name", data)
```
nit: I'd put name=data in line 63 - and remove lines 48&49.

cyfra commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,132 @@
+  """Frames Labeled In Cinema (FLIC)"""
+
+  BUILDER_CONFIGS = _make_builder_configs()
+  VERSION = tfds.core.Version("2.0.0")
```
can be removed (as you have versions in your configs)

cyfra commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,132 @@
+    extract_path = dl_manager.download_and_extract(
+        self.builder_config.url)
+    
+    data = scipy.io.loadmat(os.path.join(extract_path,
```
please use tfds.core.lazy_imports.scipy.io.loadmat and remove the direct import of scipy.io

cyfra commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,132 @@
+        self.builder_config.url)
+    
+    data = scipy.io.loadmat(os.path.join(extract_path,
+                                         self.builder_config.dir, "examples.mat"),
```
Please change the code to use tf.io.gfile to read the data (this makes the code work on non-local filesystems too):

    with tf.io.gfile.GFile(filepath, "rb") as f:
      data = tfds.core.lazy_imports.scipy.io.loadmat(f)

cyfra commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,132 @@
+        gen_kwargs={
+          "extract_path": extract_path,
+          "data": data,
+          "istrain": True,
```
nit:
having two separate bools, that are never both 'true' or both 'false' is not a great design.
Consider passing an enum or simply the 'name' argument instead.

cyfra commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,132 @@
+  def _generate_examples(self, extract_path, data, istrain, istest):
+    """Yields examples."""
+    for u_id, example in enumerate(data["examples"]):
+      if (example[7] and istrain) or (example[8] and istest):
```
avoid 'magic numbers' - what is 7? what is 8?

WilliamHYZhang commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,132 @@
+
+_DATA_OPTIONS = ["small", "full"]
+
+_URL_SUBSET = "https://drive.google.com/uc?id=0B4K3PZp8xXDJN0Fpb0piVjQ3Y3M&export=download"
```
They are the official links the dataset has provided.

WilliamHYZhang commented:
>nit:
having two separate bools, that are never both 'true' or both 'false' is not a great design.
Consider passing an enum or simply the 'name' argument instead.

>avoid 'magic numbers' - what is 7? what is 8?

I agree with this, however the dataset itself is structured like this, where for each example the 7th and 8th index respectively are booleans, `istrain` and `istest`. I was trying to reflect this in the splits.

WilliamHYZhang commented:
Thank you @cyfra for the time taken to review and your comments, appreciated very much.

cyfra commented:
> I agree with this, however the dataset itself is structured like this, where for each example the 7th and 8th index respectively are booleans, `istrain` and `istest`. I was trying to reflect this in the splits.


What you could do: is pass the 'int' argument (for example 'selection_column').
And then in the code you do:

if example[selection_column]:
  ....

WilliamHYZhang commented:
@ updated to reflect your suggestion.

WilliamHYZhang commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,134 @@
+    
+    mat_path = os.path.join(extract_path, self.builder_config.dir, "examples.mat")
+    with tf.io.gfile.GFile(mat_path) as f:
+    data = tfds.core.lazy_imports.scipy.io.loadmat(f, struct_as_record=True,
```
Sorry, didn't catch that, should pass now.

Conchylicultor commented on tensorflow_datasets/image/flic.py:
```diff
@@ -0,0 +1,134 @@
+        gen_kwargs={
+          "extract_path": extract_path,
+          "data": data,
+          "selection_column": 8,
```
Could you add a comment about what this selection_column represent ?

WilliamHYZhang commented:
Sure, no problem.

On Tue, Jan 14, 2020 at 2:56 PM Conchylicultor <notifications@github.com>
wrote:

> *@Conchylicultor* commented on this pull request.
> ------------------------------
>
> In tensorflow_datasets/image/flic.py
> <https://github.com/tensorflow/datasets/pull/1290#discussion_r366542610>:
>
> > +
> +    return [
> +      tfds.core.SplitGenerator(
> +        name=tfds.Split.TRAIN,
> +        gen_kwargs={
> +          "extract_path": extract_path,
> +          "data": data,
> +          "selection_column": 7,
> +        },
> +      ),
> +      tfds.core.SplitGenerator(
> +        name=tfds.Split.TEST,
> +        gen_kwargs={
> +          "extract_path": extract_path,
> +          "data": data,
> +          "selection_column": 8,
>
> Could you add a comment about what this selection_column represent ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/datasets/pull/1290?email_source=notifications&email_token=AJQVBJBLZBVDUNALD5IDKWTQ5YKGPA5CNFSM4JZHKWH2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRXLKRA#pullrequestreview-342799684>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJQVBJGK7K77AFVZDTGXAR3Q5YKGPANCNFSM4JZHKWHQ>
> .
>

Conchylicultor commented:
Thanks, almost good to go. 

I think for consistency with other datasets, torsobox should be a `BBoxFeature()` instead of a `Tensor(shape=(4,))`, otherwise it's not clear for the user what are the coordinate order `(x, y, xmin, ymin)` ? `(x, xmin, y, ymin)`. What do you think ?

WilliamHYZhang commented:
Oh yeah, good point. Thanks for the suggestion.

WilliamHYZhang commented:
@Conchylicultor Implemented BBox attribute, fixed some bugs, uploaded checksums.
